plat/rockchip/rk3399/drivers/pmu/pmu.c

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (c) 2016, ARM Limited and Contributors. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <arch_helpers.h>
+#include <assert.h>
+#include <bakery_lock.h>
+#include <debug.h>
+#include <delay_timer.h>
+#include <errno.h>
+#include <mmio.h>
+#include <platform.h>
+#include <platform_def.h>
+#include <plat_private.h>
+#include <pmu.h>
+#include <pmu_sram.h>
+#include <rk3399_def.h>
+#include <soc.h>
+
+static struct psram_data_t *psram_sleep_cfg =
+	(struct psram_data_t *)PSRAM_DT_BASE;
+
+/* These are wrapper macros to the Coherent Memory Bakery Lock API */
+DEFINE_BAKERY_LOCK(rockchip_lock_core_pm[PLATFORM_CORE_COUNT]);
+
+#define rockchip_core_pm_lock_get(cpu_id) \
+	bakery_lock_get(&rockchip_lock_core_pm[cpu_id])
+#define rockchip_core_pm_lock_release(cpu_id) \
+	bakery_lock_release(&rockchip_lock_core_pm[cpu_id])
+
+static uint32_t core_pm_cfg_info[PLATFORM_CORE_COUNT]
```

Comment:
There are two ways to power on or off the core power domain.
1) Control it power domain into on or off by PMU_PWRDN_CON reg, it is core_pwr_pd mode
2) Enable the core power manage in PMU_CORE_PM_CON reg,
then, if the core enter into wfi, it power domain will be
powered off automatically. it is core_pwr_wfi or core_pwr_wfi_int mode
So we need core_pm_cfg_info to distinguish which method was used now.